File: TripKit/Classes/server/parsing/TKRoutingParser.m
//
//  TKRoutingParser.m
//  TripKit
//
//  Created by Adrian Schoenig on 7/04/2015.
//  Copyright (c) 2015 SkedGo Pty Ltd. All rights reserved.
//

#import "TKRoutingParser.h"

@import CoreData;

#import <TripKit/TripKit-Swift.h>

@interface TKRoutingParser ()

@property (nonatomic, strong) NSManagedObjectContext *context;

@end

@implementation TKRoutingParser

- (id)initWithTripKitContext:(NSManagedObjectContext *)context
{
  self = [super init];
  if (self) {
    _context = context;
  }
  return self;
}

#pragma mark - Public method

- (TripRequest *)parseAndAddResultBlocking:(NSDictionary *)json
{
  __block TripRequest *result = nil;
  
  [self.context performBlockAndWait:^{
    // create an empty request
    TripRequest *request = [TripRequest insertEmptyIntoContext:self.context];
    
    // parse everything
    NSArray *added = [self parseAndAddResult:json
                                  forRequest:request
                                 orTripGroup:nil
                                orUpdateTrip:nil
                allowDuplicatingExistingTrip:YES
                                  visibility:TKTripGroupVisibilityFull];
    if (added.count == 0) {
      return;
    }
    
    BOOL success = [TKRoutingParser populate:request start:nil end:nil leaveAfter:nil arriveBy:nil queryJSON:json[@"query"]];
    if (! success) {
      return;
    }
    result = request;
  }];
  return result;
}

- (void)parseAndAddResult:(NSDictionary *)json
               completion:(void (^)(TripRequest * _Nullable request))completion
{
  [self.context performBlock:^{
    TripRequest *request = [TripRequest insertEmptyIntoContext:self.context];
    
    NSArray *added = [self parseAndAddResult:json
                                  forRequest:request
                                 orTripGroup:nil
                                orUpdateTrip:nil
                allowDuplicatingExistingTrip:YES
                                  visibility:TKTripGroupVisibilityFull];
    if (added.count == 0) {
      [self.context deleteObject:request];
      [TKLog warn:@"TKRoutingParser" text:[NSString stringWithFormat:@"Error parsing request: %@", json]];
      completion(nil);
      return;
    }
    
    BOOL success = [TKRoutingParser populate:request start:nil end:nil leaveAfter:nil arriveBy:nil queryJSON:json[@"query"]];
    if (! success) {
      [self.context deleteObject:request];
      [TKLog info:@"TKRoutingParser" text:[NSString stringWithFormat:@"Got trip without a segment from JSON: %@", json]];
      completion(nil);
      return;
    }
    completion(request);
  }];
}

- (void)parseAndAddResult:(NSDictionary *)json
            intoTripGroup:(TripGroup *)tripGroup
                  merging:(BOOL)mergeWithExistingTrips
               completion:(void (^)(NSArray<Trip *> *addedTrips))completion
{
  [self.context performBlock:^{
    TripGroup *group = (TripGroup *)[self.context objectWithID:tripGroup.objectID];
    
    NSArray *result = [self parseAndAddResult:json
                                   forRequest:nil
                                  orTripGroup:group
                                 orUpdateTrip:nil
                 allowDuplicatingExistingTrip:!mergeWithExistingTrips
                                   visibility:TKTripGroupVisibilityFull];
    completion(result);
  }];
}

- (void)parseAndAddResult:(NSDictionary *)json
               forRequest:(TripRequest *)request
                  merging:(BOOL)mergeWithExistingTrips
               completion:(void (^)(NSArray<Trip *> *addedTrips))completion
{
  [self.context performBlock:^{
    NSArray *result = [self parseAndAddResult:json
                                   forRequest:request
                                  orTripGroup:nil
                                 orUpdateTrip:nil
                 allowDuplicatingExistingTrip:!mergeWithExistingTrips
                                   visibility:TKTripGroupVisibilityFull];
    completion(result);
  }];
}

- (NSArray<Trip *>*)parseAndAddResult:(NSDictionary *)json
                           forRequest:(TripRequest *)request
                              merging:(BOOL)mergeWithExistingTrips
                           visibility:(TKTripGroupVisibility)visibility
{
  return [self parseAndAddResult:json
                      forRequest:request
                     orTripGroup:nil
                    orUpdateTrip:nil
    allowDuplicatingExistingTrip:!mergeWithExistingTrips
                      visibility:visibility];
}

- (void)parseAndAddResult:(NSDictionary *)keyToTripGroups
     withSegmentTemplates:(NSArray *)segmentTemplatesJson
                andAlerts:(nullable NSArray *)alertJson
               completion:(void (^)(NSDictionary *keyToAddedTrips))completion
{
  [self.context performBlock:^{
    // create an empty request
    NSMutableDictionary *keyToTrips = [NSMutableDictionary dictionaryWithCapacity:keyToTripGroups.count];
    
    [keyToTripGroups enumerateKeysAndObjectsUsingBlock:
     ^(id<NSCopying> key, NSArray *tripGroupsArray, BOOL *stop) {
#pragma unused(stop)
       TripRequest *request = [TripRequest insertEmptyIntoContext:self.context];
       
       NSArray *newTrips = [self parseAndAddResultWithTripGroups:tripGroupsArray
                                                segmentTemplates:segmentTemplatesJson
                                                          alerts:alertJson
                                                      forRequest:request
                                                     orTripGroup:nil
                                                    orUpdateTrip:nil
                                    allowDuplicatingExistingTrip:YES
                                                      visibility:TKTripGroupVisibilityFull];
       if (newTrips.count > 0) {
         keyToTrips[key] = newTrips;
       }
     }];
    completion(keyToTrips);
  }];
}

- (void)parseJSON:(NSDictionary *)json
     updatingTrip:(Trip *)trip
       completion:(void (^)(Trip *updatedTrip))completion
{
  [self.context performBlock:^{
    [self parseAndAddResult:json
                 forRequest:nil
                orTripGroup:nil
               orUpdateTrip:trip
allowDuplicatingExistingTrip:YES // we don't actually create a duplicate
                 visibility:TKTripGroupVisibilityFull];
    completion(trip);
  }];
}


#pragma mark - Private helpers


- (NSArray *)parseAndAddResult:(NSDictionary *)json
                    forRequest:(nullable TripRequest *)request
                   orTripGroup:(nullable TripGroup *)insertIntoGroup
                  orUpdateTrip:(nullable Trip *)tripToUpdate
  allowDuplicatingExistingTrip:(BOOL)allowDuplicates
                    visibility:(TKTripGroupVisibility)visibility

{
  NSString *error = json[@"error"];
  if (error) {
    [TKLog warn:@"TKRoutingParser" text:[NSString stringWithFormat:@"Error while parsing: %@", error]];
    return @[];
  }
  
  NSArray *segmentTemplatesArray = json[@"segmentTemplates"];
  NSArray *alertsArray = json[@"alerts"];
  NSArray *tripGroupsArray = json[@"groups"];
  return [self parseAndAddResultWithTripGroups:tripGroupsArray
                              segmentTemplates:segmentTemplatesArray
                                        alerts:alertsArray
                                    forRequest:request
                                   orTripGroup:insertIntoGroup
                                  orUpdateTrip:tripToUpdate
                  allowDuplicatingExistingTrip:allowDuplicates
                                    visibility:visibility];
}

- (NSArray *)parseAndAddResultWithTripGroups:(NSArray *)tripGroupsArray
                            segmentTemplates:(NSArray *)segmentTemplatesArray
                                      alerts:(NSArray *)alertsArray
                                  forRequest:(nullable TripRequest *)request
                                 orTripGroup:(nullable TripGroup *)insertIntoGroup
                                orUpdateTrip:(nullable Trip *)tripToUpdate
                allowDuplicatingExistingTrip:(BOOL)allowDuplicates
                                  visibility:(TKTripGroupVisibility)visibility
{
  ZAssert(self.context, @"Managed object context required!");
    
  // let's check if the request is still alive
  if (! request) {
    request = insertIntoGroup.request;
  }
  if (! request) {
    request = tripToUpdate.request;
  }
  if (! request.managedObjectContext || [request isDeleted]) {
    return @[];
  }
  
  TKTripGroupVisibility updateTripVisibility = tripToUpdate.tripGroup.visibility;
  
  // At first, we need to parse the segment templates, since the trips will reference them
  NSMutableDictionary *segmentHashToTemplateDictionaryDict = [NSMutableDictionary dictionaryWithCapacity:segmentTemplatesArray.count];
  NSMutableSet *addedTemplateHashCodes = [NSMutableSet setWithCapacity:segmentTemplatesArray.count];
  for (NSDictionary *segmentTemplateDict in segmentTemplatesArray) {
    NSNumber *hashCode = segmentTemplateDict[@"hashCode"];
    [segmentHashToTemplateDictionaryDict setValue:segmentTemplateDict forKey:[hashCode description]];

    BOOL hashCodeExists = [SegmentTemplate segmentTemplateHashCode:hashCode.integerValue
                                            existsInTripKitContext:self.context];
    if (hashCodeExists) {
      [addedTemplateHashCodes addObject:hashCode];
    }
  }
  
  // Next we parse the alerts
  [TKAPIToCoreDataConverter updateOrAddAlerts:alertsArray
                             inTripKitContext:self.context];
  
  // Now parse the groups
  NSSet *previousTrips = [request trips];
  NSMutableArray *tripsToReturn = [NSMutableArray array];
  for (NSDictionary * tripGroupDict in tripGroupsArray) {
    NSMutableSet *newTrips = [NSMutableSet set];
    
    // iterate over routes
    for (NSDictionary * tripDict in tripGroupDict[@"trips"]) {
      
      // ignore "nothing" results
      if ([tripDict[@"accuracy"] isEqualToString:@"nothing"])
        continue;
      
      // create route item
      Trip *trip;
      if (tripToUpdate) {
        trip = tripToUpdate;
      } else {
        trip = [NSEntityDescription insertNewObjectForEntityForName:NSStringFromClass([Trip class]) inManagedObjectContext:self.context];
      }
      trip.arrivalTime   = [TKParserHelper parseDate:tripDict[@"arrive"]];;
      trip.departureTime = [TKParserHelper parseDate:tripDict[@"depart"]];;
      
      // update values if we received them, otherwise keep old
      trip.mainSegmentHashCode  = [tripDict[@"mainSegmentHashCode"] intValue];
      trip.totalCalories        = [tripDict[@"caloriesCost"] floatValue];
      trip.totalCarbon          = [tripDict[@"carbonCost"] floatValue];
      trip.totalHassle          = [tripDict[@"hassleCost"] floatValue];
      trip.totalScore           = [tripDict[@"weightedScore"] floatValue];
      trip.totalPrice           = tripDict[@"moneyCost"]            ?: trip.totalPrice;
      trip.totalPriceUSD        = tripDict[@"moneyUSDCost"]         ?: trip.totalPriceUSD;
      trip.currencyCode         = tripDict[@"currency"]             ?: trip.currencyCode;
      trip.budgetPoints         = tripDict[@"budgetPoints"]         ?: trip.budgetPoints;
      trip.saveURLString        = tripDict[@"saveURL"]              ?: trip.saveURLString;
      trip.shareURLString       = tripDict[@"shareURL"]             ?: trip.shareURLString;
      trip.temporaryURLString   = tripDict[@"temporaryURL"]         ?: trip.temporaryURLString;
      trip.updateURLString      = tripDict[@"updateURL"]            ?: trip.updateURLString;
      trip.progressURLString    = tripDict[@"progressURL"]          ?: trip.progressURLString;
      trip.plannedURLString     = tripDict[@"plannedURL"]           ?: trip.plannedURLString;
      trip.logURLString         = tripDict[@"logURL"]               ?: trip.logURLString;

      if ([tripDict[@"availability"] isKindOfClass:[NSString class]]) {
        trip.missedBookingWindow = [@"MISSED_PREBOOKING_WINDOW" isEqualToString:tripDict[@"availability"]];
        trip.isCanceled = [@"CANCELLED" isEqualToString:tripDict[@"availability"]];
      }
      
      if ([tripDict[@"bundleId"] isKindOfClass:[NSString class]]) {
        trip.bundleId = tripDict[@"bundleId"];
      }
      
      [trip calculateDuration];
      
      // updated trip isn't strictly speaking new, but we want to process it as a successful match.
      [newTrips addObject:trip];
      
      NSMutableSet *unmatchedSegmentReferences = nil;
      if (tripToUpdate) {
        unmatchedSegmentReferences = [tripToUpdate.segmentReferences mutableCopy];
      }
      
      int segmentCount = 0;
      for (NSDictionary *refDict in tripDict[@"segments"]) {
        // create the reference object
        SegmentReference *reference = nil;
        NSNumber *hashCode = refDict[@"segmentTemplateHashCode"];
        ZAssert(hashCode != nil, @"No hash code in %@", refDict);
        NSDictionary *templateDict = segmentHashToTemplateDictionaryDict[[hashCode description]];
        ZAssert(templateDict != nil, @"Missing template for %@", hashCode);
        BOOL isNewTemplate = ![addedTemplateHashCodes containsObject:hashCode];
        
        if (tripToUpdate) {
          for (SegmentReference *existingReference in unmatchedSegmentReferences) {
            if ([existingReference.templateHashCode isEqualToNumber:hashCode]) {
              reference = existingReference;
              break;
            }
          }
          if (reference) {
            [unmatchedSegmentReferences removeObject:reference];
          } else {
            [trip clearSegmentCaches];
          }
        }
        
        if (!reference
            && (isNewTemplate || YES == [SegmentTemplate segmentTemplateHashCode:hashCode.integerValue existsInTripKitContext:self.context])) {
          reference = [NSEntityDescription insertNewObjectForEntityForName:NSStringFromClass([SegmentReference class]) inManagedObjectContext:self.context];
        }
        
        if (! reference) {
          continue;
        }
        
        [reference _populateFromDictionary:refDict];
        
        Service *service = nil;
        NSString *serviceCode = refDict[@"serviceTripID"];
        if (serviceCode) {
          // public-transport
          
          // create a service object if necessary
          service = [Service fetchOrInsertServiceWithCode:serviceCode
                                         inTripKitContext:self.context];
          
          // always update these as those might be new or updated, as long as they didn't get deleted
          TKColor *newColor = [TKParserHelper colorForDictionary:refDict[@"serviceColor"]];
          service.color     = newColor                      ?: service.color;
          service.frequency = refDict[@"frequency"]         ?: service.frequency;
          service.lineName  = refDict[@"serviceName"]       ?: service.lineName;
          service.direction = refDict[@"serviceDirection"]  ?: service.direction;
          service.number		= refDict[@"serviceNumber"]     ?: service.number;
          reference.service = service;
          
          reference.bicycleAccessible = [refDict[@"bicycleAccessible"] boolValue];

          [reference _setWheelchairAccessibility:refDict[@"wheelchairAccessible"]];
          
          // set the trip status
          if (service.frequency.integerValue == 0) {
            trip.departureTimeIsFixed = YES;
          }
          
          // update the real-time status
          NSString *realTimeStatus = refDict[@"realTimeStatus"];
          [TKCoreDataParserHelper adjustService:service forRealTimeStatusString:realTimeStatus];
          
          // keep the vehicles
          [TKAPIToCoreDataConverter updateVehiclesForService:service
                                              primaryVehicle:refDict[@"realtimeVehicle"]
                                         alternativeVehicles:refDict[@"realtimeVehicleAlternatives"]];
          
        } else {
          // private transport
          [TKCoreDataParserHelper updateVehiclesForSegmentReference:reference
                                             primaryVehicle:refDict[@"realtimeVehicle"]
                                        alternativeVehicles:nil];
        }

        reference.templateHashCode = hashCode;
        reference.startTime = [TKParserHelper parseDate:refDict[@"startTime"]];;
        reference.endTime = [TKParserHelper parseDate:refDict[@"endTime"]];;
        reference.timesAreRealTime = [refDict[@"realTime"] boolValue];

        reference.alertHashCodes = refDict[@"alertHashCodes"];
        
        
        ZAssert(templateDict, @"No segment template found for code %@", hashCode);
        if (isNewTemplate) {
          [SegmentTemplate insertNewTemplateFromDictionary:templateDict forService:service relativeTime:reference.startTime intoContext:self.context];
          [addedTemplateHashCodes addObject:hashCode];
        } else if (service) {
          // We don't need to insert the full template, but need to add
          // shapes for that service
          
          TKModeInfo *modeInfo = [service modeInfo]; // *not* using `findModeInfo`
            // as we might have just created this, and it'll get populated from
            // templateDict
          
          [SegmentTemplate insertNewShapesFromDictionary:templateDict forService:service relativeTime:reference.startTime modeInfo:modeInfo intoContext:self.context];
        }
        
        reference.index = @(segmentCount++);
        reference.trip = trip;
      }
      
      // Clean-up unmatched references
      if (unmatchedSegmentReferences.count > 0) {
        for (SegmentReference *unmatched in unmatchedSegmentReferences) {
          [self.context deleteObject:unmatched];
        }
      }
      ZAssert(trip.segmentReferences.count > 0, @"Trip has no segments: %@", trip);
    }
    
    if (newTrips.count > 0) {
      TripGroup *tripGroup = nil;
      NSMutableArray *addedTrips = [NSMutableArray arrayWithCapacity:newTrips.count];
      
      // Check if we already have a similar route. If so, we'll add ALL routes of this group to the group of an existing route.
      for (Trip *trip in newTrips) {
        // let's check if the request is still alive
        if (nil == request.managedObjectContext || [request isDeleted]) {
          return nil;
        }
        
        Trip *existingNearlyIdenticalTrip = nil;
        if (!allowDuplicates) {
          existingNearlyIdenticalTrip = [Trip findSimilarTripTo:trip inList:previousTrips];
        }
        
        if (! existingNearlyIdenticalTrip) {
          [addedTrips addObject:trip];
          [tripsToReturn addObject:trip];
        } else if (existingNearlyIdenticalTrip.tripGroup) {
          // remember route group, but don't add this
          tripGroup = existingNearlyIdenticalTrip.tripGroup;
          [tripsToReturn addObject:existingNearlyIdenticalTrip];
          
          // delete this
          [self.context deleteObject:trip];
        }
      }
      
      if (insertIntoGroup) {
        tripGroup = insertIntoGroup;
      } else if (tripToUpdate) {
        tripGroup = tripToUpdate.tripGroup;
      }
      
      if (addedTrips.count > 0) {
        if (nil == tripGroup) {
          // create a new route group
          tripGroup = [NSEntityDescription insertNewObjectForEntityForName:NSStringFromClass([TripGroup class]) inManagedObjectContext:self.context];
          tripGroup.request = request;
        }
        
        tripGroup.visibility = visibility;
        
        // always update frequency + sources (if there are any)
        tripGroup.frequency = tripGroupDict[@"frequency"] ?: tripGroup.frequency;
        tripGroup.sourcesRaw = tripGroupDict[@"sources"] ?: tripGroup.sourcesRaw;
        
        for (Trip *trip in newTrips) {
          if (trip.managedObjectContext != nil) {
            trip.tripGroup = tripGroup;
          }
        }
        if (!tripToUpdate) {
          [tripGroup adjustVisibleTrip];
        }
      }
    }
  }
  
  // restore visibility
  if (tripToUpdate) {
    tripToUpdate.tripGroup.visibility = updateTripVisibility;
  }
  return tripsToReturn;
}

@end

File: Sources/TripKit/server/parsing/TKRoutingParser.swift
//
//  TKRoutingParser.swift
//  TripKit
//
//  Created by Adrian Schönig on 10/8/21.
//  Copyright © 2021 SkedGo Pty Ltd. All rights reserved.
//

import Foundation

import CoreData

public final class TKRoutingParser {
  
  private init() {}
  
  enum ParserError: Error {
    case didNotFinish
    case mismatchingResponse
    case serverError(String)
  }
  
  public static func add(_ response: TKAPI.RoutingResponse, into context: NSManagedObjectContext = TripKit.shared.tripKitContext, completion: @escaping (Result<TripRequest, Error>) -> Void) {
    context.perform {
      let request = TripRequest(context: context)
      request.timeCreated = .init()
      do {
        try add(response,
                mode: .addTo(request),
                context: context,
                allowDuplicates: true,
                visibility: .full
        )
        
        Self.populate(request, using: response.query)
        completion(.success(request))
        
      } catch {
        context.delete(request)
        completion(.failure(error))
      }
    }
  }
  
  static func addBlocking(_ response: TKAPI.RoutingResponse, into context: NSManagedObjectContext) throws -> TripRequest {
    var result: Result<TripRequest, Error>? = nil
    context.performAndWait {
      let request = TripRequest(context: context)
      request.timeCreated = .init()
      result = Result {
        try add(response,
                mode: .addTo(request),
                context: context,
                allowDuplicates: true,
                visibility: .full
        )
        
        Self.populate(request, using: response.query)
        return request
      }
    }
    return try result.orThrow(ParserError.didNotFinish).get()
  }
  
  static func add(_ response: TKAPI.RoutingResponse, to request: TripRequest, merge: Bool, visibility: TripGroup.Visibility = .full, completion: @escaping (Result<[Trip], Error>) -> Void) {
    guard let context = request.managedObjectContext else {
      assertionFailure("Request's context went missing.")
      completion(.success([]))
      return
    }
    
    context.perform {
      do {
        let added = try add(response,
                mode: .addTo(request),
                context: context,
                allowDuplicates: !merge,
                visibility: visibility
        )
        completion(.success(added))
        
      } catch {
        completion(.failure(error))
      }
    }
  }
  
  static func add(_ response: TKAPI.RoutingResponse, to group: TripGroup, merge: Bool, completion: @escaping (Result<[Trip], Error>) -> Void) {
    guard let context = group.managedObjectContext else {
      assertionFailure("Trip group's context went missing.")
      completion(.success([]))
      return
    }
    
    context.perform {
      do {
        let added = try add(response,
                mode: .appendTo(group),
                context: context,
                allowDuplicates: !merge,
                visibility: .full
        )
        completion(.success(added))
        
      } catch {
        completion(.failure(error))
      }
    }
  }
  
  static func update(_ trip: Trip, from response: TKAPI.RoutingResponse, completion: @escaping (Result<Trip, Error>) -> Void) {
    guard let context = trip.managedObjectContext else {
      assertionFailure("Trip's context went missing.")
      completion(.success(trip))
      return
    }
    
    context.perform {
      completion(Result {
        try add(response,
            mode: .update(trip),
            context: context,
            allowDuplicates: true, // we don't actually create a duplicate
            visibility: .full
        )
        try context.save()
        return trip
      })
    }
  }
  
  public static func add<Key>(groups: [Key: [TKAPI.TripGroup]], templates: [TKAPI.SegmentTemplate], alerts: [TKAPI.Alert], into context: NSManagedObjectContext = TripKit.shared.tripKitContext, completion: @escaping (Result<[Key: [Trip]], Error>) -> Void) {
    context.perform {
      do {
        var keyToTrips: [Key: [Trip]] = [:]
        for (key, groups) in groups {
          let request = TripRequest(context: context)
          request.timeCreated = .init()
          let trips = try add(groups: groups, templates: templates, alerts: alerts, mode: .addTo(request), context: context, allowDuplicates: true, visibility: .full)
          if !trips.isEmpty {
            keyToTrips[key] = trips
          }
        }
        completion(.success(keyToTrips))
      } catch {
        completion(.failure(error))
      }
    }
  }

  @discardableResult
  private static func add(
    _ response: TKAPI.RoutingResponse,
    mode: ParseMode,
    context: NSManagedObjectContext,
    allowDuplicates: Bool,
    visibility: TripGroup.Visibility
  ) throws -> [Trip] {
    
    if let error = response.error {
      throw ParserError.serverError(error)
    }
    
    guard
      let groups = response.groups,
      let templates = response.segmentTemplates
    else {
      throw ParserError.mismatchingResponse
    }
    
    return try add(groups: groups, templates: templates, alerts: response.alerts, mode: mode, context: context, allowDuplicates: allowDuplicates, visibility: visibility)
  }

  private enum ParseMode {
    case addTo(TripRequest)
    case appendTo(TripGroup)
    case update(Trip)
    
    var request: TripRequest {
      switch self {
      case .addTo(let request): return request
      case .appendTo(let group): return group.request
      case .update(let trip): return trip.request
      }
    }
    
    var isUpdate: Bool {
      switch self {
      case .update: return true
      case .addTo, .appendTo: return false
      }
    }
  }
  
  private static func add(
    groups: [TKAPI.TripGroup],
    templates: [TKAPI.SegmentTemplate],
    alerts: [TKAPI.Alert],
    mode: ParseMode,
    context: NSManagedObjectContext,
    allowDuplicates: Bool,
    visibility: TripGroup.Visibility
  ) throws -> [Trip] {
    
    let request = mode.request
    guard request.managedObjectContext != nil, !request.isDeleted else {
      return []
    }

    // We want to maintain these and reset to it at the very end
    let originalVisibility: TripGroup.Visibility
    switch mode {
    case .update(let trip): originalVisibility = trip.tripGroup.visibility
    default: originalVisibility = .full
    }
    
    // At first, we need to parse the segment templates, since the trips will reference them
    var templateByHashCodes: [Int: TKAPI.SegmentTemplate] = [:]
    var existingTemplateHashCodes = Set<Int>()
    for apiTemplate in templates {
      templateByHashCodes[apiTemplate.hashCode] = apiTemplate
      if SegmentTemplate.segmentTemplate(withHashCode: apiTemplate.hashCode, existsIn: context) {
        existingTemplateHashCodes.insert(apiTemplate.hashCode)
      }
    }
    
    // Persist the alerts
    TKAPIToCoreDataConverter.updateOrAddAlerts(alerts, in: context)
    
    // Now we create the trips
    let previousTrips = request.trips
    var tripsToReturn: [Trip] = []
    for apiGroup in groups {
      var newTrips = Set<Trip>()
      
      for apiTrip in apiGroup.trips {
        let trip: Trip
        var unmatchedSegmentReferencesByHashCode: [Int: SegmentReference] = [:]
        switch mode {
        case .update(let existing):
          trip = existing
          if let references = existing.segmentReferences {
            unmatchedSegmentReferencesByHashCode = Dictionary(uniqueKeysWithValues: references.map { ($0.templateHashCode?.intValue ?? 0, $0) })
          }
        case .addTo, .appendTo:
          trip = Trip(context: context)
        }
        
        trip.departureTime = apiTrip.depart
        trip.arrivalTime = apiTrip.arrive
        trip.calculateDuration()
        
        trip.mainSegmentHashCode = Int32(apiTrip.mainSegmentHashCode)
        trip.totalCalories = Float(apiTrip.caloriesCost)
        trip.totalCarbon = Float(apiTrip.carbonCost)
        trip.totalHassle = Float(apiTrip.hassleCost)
        trip.totalScore = Float(apiTrip.weightedScore)
        
        // fallback to old values (if we're updating)
        trip.totalPrice = apiTrip.moneyCost.map(NSNumber.init) ?? trip.totalPrice
        trip.update(\.totalPrice, value: apiTrip.moneyCost)
        trip.update(\.totalPriceUSD, value: apiTrip.moneyCostUSD)
        trip.update(\.currencyCode, value: apiTrip.currency)
        trip.update(\.saveURLString, value: apiTrip.saveURL?.absoluteString)
        trip.update(\.shareURLString, value: apiTrip.shareURL?.absoluteString)
        trip.update(\.temporaryURLString, value: apiTrip.temporaryURL?.absoluteString)
        trip.update(\.updateURLString, value: apiTrip.updateURL?.absoluteString)
        trip.update(\.progressURLString, value: apiTrip.progressURL?.absoluteString)
        trip.update(\.plannedURLString, value: apiTrip.plannedURL?.absoluteString)
        trip.update(\.logURLString, value: apiTrip.logURL?.absoluteString)
        trip.update(\.bundleId, value: apiTrip.bundleId)
        
        switch apiTrip.availability {
        case .missedPrebookingWindow: trip.missedBookingWindow = true
        case .canceled: trip.isCanceled = true
        case .none: break
        }
        
        // updated trip isn't strictly speaking new, but we want to process it
        // as a successful match.
        newTrips.insert(trip)
        
        for (index, apiReference) in apiTrip.segments.enumerated() {
          let hashCode = apiReference.segmentTemplateHashCode
          guard let apiTemplate = templateByHashCodes[hashCode] else {
            assertionFailure("Missing template for \(hashCode)")
            continue
          }
          let isNewTemplate = !existingTemplateHashCodes.contains(hashCode)
          
          var reference: SegmentReference! = nil
          if mode.isUpdate {
            if let existing = unmatchedSegmentReferencesByHashCode.removeValue(forKey: hashCode)  {
              reference = existing
            } else {
              trip.clearSegmentCaches()
            }
          }
          if reference == nil, (isNewTemplate || SegmentTemplate.segmentTemplate(withHashCode: hashCode, existsIn: context)) {
            reference = SegmentReference(context: context)
          }
          guard let reference = reference else {
            continue
          }
          
          reference.templateHashCode = NSNumber(value: hashCode)
          reference.startTime = apiReference.startTime
          reference.endTime = apiReference.endTime
          reference.timesAreRealTime = apiReference.timesAreRealTime
          reference.alertHashCodes = apiReference.alertHashCodes.map(NSNumber.init)
          reference.populate(from: apiReference)
          
          let maybeService: Service?
          if let tripID = apiReference.serviceTripID {
            // Public transport
            let service = Service.fetchOrInsert(code: tripID, in: context)
            maybeService = service
            
            // always update these as those might be new or updated, as long as they didn't get deleted
            service.color = apiReference.serviceColor?.color ?? service.color
            service.frequency = apiReference.frequency.map(NSNumber.init) ?? service.frequency
            service.lineName = apiReference.lineName ?? service.lineName
            service.direction = apiReference.direction ?? service.direction
            service.number = apiReference.number ?? service.number
            reference.service = service
            
            reference.bicycleAccessible = apiReference.bicycleAccessible
            reference.wheelchairAccessibility = TKWheelchairAccessibility(bool: apiReference.wheelchairAccessible)
            
            // If we have any time-tabled service, the whole trip is timetabled
            if service.frequency == nil {
              trip.departureTimeIsFixed = true
            }
            
            service.adjustRealTimeStatus(for: apiReference.realTimeStatus ?? .incapable)
            service.addVehicles(primary: apiReference.realTimeVehicle, alternatives: apiReference.realTimeVehicleAlternatives)
            
          } else {
            // Non-PT
            if let apiVehicle = apiReference.realTimeVehicle {
              if let existing = reference.realTimeVehicle {
                existing.update(with: apiVehicle)
              } else {
                reference.realTimeVehicle = Vehicle(from: apiVehicle, into: context)
              }
            }
            maybeService = nil
          }
          
          // Finally, insert the template that this reference points to
          if isNewTemplate {
            SegmentTemplate.insertNewTemplate(from: apiTemplate, for: maybeService, relativeTime: apiReference.startTime, into: context)
            existingTemplateHashCodes.insert(hashCode)
          } else if let service = maybeService {
            // We don't need to insert the full template, but need to add
            // shapes for that service
            
            let modeInfo = service.modeInfo // *not* using `findModeInfo` as we
                        // might have just created this, and it'll get populated
                        // from `apiTemplate`
            SegmentTemplate.insertNewShapes(from: apiTemplate, for: service, relativeTime: reference.startTime, modeInfo: modeInfo, context: context)
          }
          
          reference.index = Int16(index)
          reference.trip = trip
        }
        
        // Clean-up unmatched references
        if !unmatchedSegmentReferencesByHashCode.isEmpty {
          _ = unmatchedSegmentReferencesByHashCode.values.map(context.delete(_:))
        }
        assert((trip.segmentReferences ?? []).count > 0, "Trip has no segments: \(trip)")
      }
      
      if !newTrips.isEmpty {
        var existingGroup: TripGroup? = nil
        var addedTrips: [Trip] = []
        
        // Check if we already have a similar route. If so, we'll add ALL routes of this group to the group of an existing route.
        for trip in newTrips {
          guard request.context != nil, !request.isDeleted else {
            return []
          }

          let existingSimilarTrip = allowDuplicates
            ? nil
            : Trip.findSimilarTrip(to: trip, in: previousTrips)
          if let similar = existingSimilarTrip {
            // remember group, but don't add it
            existingGroup = similar.tripGroup
            tripsToReturn.append(similar)
            if similar != trip {
              context.delete(trip)
            }
          } else {
            addedTrips.append(trip)
            tripsToReturn.append(trip)
          }
        }
        
        switch mode {
        case .appendTo(let group): existingGroup = group
        case .update(let trip): existingGroup = trip.tripGroup
        case .addTo: break
        }
        
        if !addedTrips.isEmpty {
          let tripGroup: TripGroup
          if let existing = existingGroup {
            tripGroup = existing
          } else {
            tripGroup = TripGroup(context: context)
            tripGroup.request = request
          }
          tripGroup.visibility = visibility
          tripGroup.frequency = apiGroup.frequency.map(NSNumber.init) ?? tripGroup.frequency
          tripGroup.sources = apiGroup.sources
          newTrips
            .filter { $0.managedObjectContext != nil }
            .forEach { $0.tripGroup = tripGroup }
          
          if !mode.isUpdate {
            tripGroup.adjustVisibleTrip()
          }
        }
      }
    }
    
    // restore visibility
    switch mode {
    case .update(let trip): trip.tripGroup.visibility = originalVisibility
    default: break
    }
    
    return tripsToReturn
  }
  
}

extension Trip {
  func update<V>(_ path: WritableKeyPath<Trip, V>, value: V?) {
    guard let value = value else { return }
    var trip = self
    trip[keyPath: path] = value
  }
  
  func update(_ path: WritableKeyPath<Trip, NSNumber?>, value: Double?) {
    guard let value = value else { return }
    var trip = self
    trip[keyPath: path] = NSNumber(value: value)
  }
}
